### PR TITLE
Sbt integration, enables remote sbt compilation

### PR DESCRIPTION
**Pull request by wpopielarski**

### CODE REVIEW

**Comment by sschaef** [comment]
please rebuild this

**Comment by wpopielarski** [comment]
rebuild this please

**Comment by sschaef** [comment]
retest this please

**Comment by sschaef** [comment]
Man, which imbecile developed the API to communicate with Jenkins. I always have to look it up

File: org.scala-ide.sbt.core/src/org/scalaide/sbt/core/SbtBuild.scala [hunk 1]
```diff
@@ -59,6 +67,22 @@ object SbtBuild extends AnyRef with HasLogger {
     }
   }
 
+  def shutdown(implicit system: ExecutionContext): Future[Boolean] = Future.sequence {
+    builds.values
+  }.flatMap { builds =>
+    buildsLock.synchronized {
+      builds.foreach { _.shutdown() }
+    }
+    @tailrec def areClosed: Boolean = if (builds.forall { _.isClosed })
```
Comment (sschaef) [inline org.scala-ide.sbt.core/src/org/scalaide/sbt/core/SbtBuild.scala]: I think this polling is too expensive. We should put in some `Thread.sleep(100)`. Or use an akka scheduler to rerun the test continually.

File: org.scala-ide.sbt.core/src/org/scalaide/sbt/core/SbtBuild.scala [hunk 2]
```diff
@@ -59,6 +67,22 @@ object SbtBuild extends AnyRef with HasLogger {
     }
   }
 
+  def shutdown(implicit system: ExecutionContext): Future[Boolean] = Future.sequence {
```
Comment (sschaef) [inline org.scala-ide.sbt.core/src/org/scalaide/sbt/core/SbtBuild.scala]: side effecting method. misses `()`

File: org.scala-ide.sbt.core/src/org/scalaide/sbt/core/SbtBuild.scala (hunk 1 above)
Comment (sschaef) [inline org.scala-ide.sbt.core/src/org/scalaide/sbt/core/SbtBuild.scala]: Also, thin method can't return `false`. Therefore there is no need to let it return anything.

File: org.scala-ide.sbt.core/src/org/scalaide/sbt/core/SbtBuild.scala [hunk 3]
```diff
@@ -153,6 +183,16 @@ class SbtBuild private (val buildRoot: File, sbtClient: RichSbtClient, console:
     implicit val kp = KeyProvider.TaskKeyKP(sbtClient.client)
     sbtClient.keyValue(projectName, keyName, config)
   }
+
+  def compilationResult(compileId: Long, buildReporter: RemoteBuildReporter) =
```
Comment (sschaef) [inline org.scala-ide.sbt.core/src/org/scalaide/sbt/core/SbtBuild.scala]: public method misses explicit return type

File: org.scala-ide.sbt.core/src/org/scalaide/sbt/core/builder/RemoteBuilder.scala [hunk 4]
```diff
@@ -1,35 +1,55 @@
...
-  override def build(kind: Int, args: JMap[String, String], monitor: IProgressMonitor): Array[IProject] = {
+  override def build(ingnoreAddedOrUpdated: Set[IFile], ignoreRemoved: Set[IFile], monitor: SubMonitor): Unit = {
     implicit val system = SbtRemotePlugin.system
     import system.dispatcher
 
-    val project = getProject()
-    val build = SbtBuild.buildFor(project.getLocation().toFile())
-    val res = build.flatMap { _.compile(project) }
+    val build = SbtBuild.buildFor(project.underlying.getLocation().toFile())
+    val res = build.flatMap { sbtBuild =>
```
Comment (sschaef) [inline org.scala-ide.sbt.core/src/org/scalaide/sbt/core/builder/RemoteBuilder.scala]: for comprehension may fit better here?

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ScalaProject.scala [hunk 5]
```diff
@@ -582,10 +585,11 @@ class ScalaProject private(val underlying: IProject) extends ClasspathManagement
       logger.info("BM: SBT enhanced Build Manager for " + IScalaPlugin().scalaVersion + " Scala library")
```
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ScalaProject.scala]: this logging should be replaced with something that tells which build manager is used.

**Comment by sschaef** [comment]
I found a bug. Following setup: Two projects `A` and `B`, where `B` depends on `A`. `A` uses sbt-remote building and `B` uses internal builder. If Eclipse is started and I invoke builder on `B`. I get the following error:

```
2016-07-29 12:45:01,219 DEBUG [Worker-9] - System.out - java.lang.InstantiationException: org.scalaide.sbt.core.builder.RemoteBuilder
	at java.lang.Class.newInstance(Class.java:427)
	at org.eclipse.core.internal.registry.osgi.RegistryStrategyOSGI.createExecutableExtension(RegistryStrategyOSGI.java:184)
	at org.eclipse.core.internal.registry.ExtensionRegistry.createExecutableExtension(ExtensionRegistry.java:905)
	at org.eclipse.core.internal.registry.ConfigurationElement.createExecutableExtension(ConfigurationElement.java:243)
	at org.eclipse.core.internal.registry.ConfigurationElementHandle.createExecutableExtension(ConfigurationElementHandle.java:55)
	at org.eclipse.core.internal.events.BuildManager.instantiateBuilder(BuildManager.java:902)
	at org.eclipse.core.internal.events.BuildManager.initializeBuilder(BuildManager.java:855)
	at org.eclipse.core.internal.events.BuildManager.getBuilder(BuildManager.java:547)
	at org.eclipse.core.internal.events.BuildManager.getBuilder(BuildManager.java:576)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:244)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:301)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:304)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:360)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:383)
	at org.eclipse.core.internal.resources.Workspace.buildInternal(Workspace.java:487)
	at org.eclipse.core.internal.resources.Workspace.build(Workspace.java:406)
	at org.eclipse.ui.actions.BuildAction$1.runInWorkspace(BuildAction.java:285)
	at org.eclipse.core.internal.resources.InternalWorkspaceJob.run(InternalWorkspaceJob.java:39)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: java.lang.NoSuchMethodException: org.scalaide.sbt.core.builder.RemoteBuilder.<init>()
	at java.lang.Class.getConstructor0(Class.java:3082)
	at java.lang.Class.newInstance(Class.java:412)
	... 19 more
```

If I however, first build `A`, then the sbt-remote server is successfully started and building works fine. Therefore, for some reason, if sbt-remote is not started the builder can't figure out that `A` is a sbt-remote project.

**Comment by sschaef** [comment]
The error log shows `compile execution N failed`. It may make sense to filter out this error message, it doesn't give the user any information.

**Comment by sschaef** [comment]
(has nothing to do with project building)

If the sbt configuration is changed, we are no longer allowed to run sbteclipse because it would create different configuration than the sbt-project-import wizard. We need to check if the sbt configuration changed since the last time sbt-remote was used and then recreate the configuration, otherwise we will see random errors (classpath may have changed for example).

**Comment by wpopielarski** [comment]
retest this please

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ScalaProject.scala [hunk 6]
```diff
@@ -589,7 +589,11 @@ class ScalaProject private(val underlying: IProject) extends ClasspathManagement
         if (!storage.getBoolean(useSbtCompilerProperty))
           new SbtScopesBuildManager(this, settings)
```
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ScalaProject.scala]: I would also make this one an instance of the extension point.

Comment (wpopielarski) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ScalaProject.scala]: ok for doing it an extension,
going back to described problems with cleaning two dependent project. now I suffer with NPE in SbtBuild.sbtClientWatcher#135. 'system' is null and means that SbtRemotePlugin.system is null in such scenario. I look at this on Monday

**Comment by wpopielarski** [comment]
@sschaef  I'm done

**Comment by sschaef** [comment]
The .classpath file needs to be changed:
```
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
modified: org.scala-ide.sbt.core/.classpath
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
@ .classpath:11 @
        <classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.6"/>
        <classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
        <classpathentry kind="src" output="target/classes" path="src"/>
        <classpathentry kind="lib" path="target/lib/akka-osgi_2.11.jar" sourcepath="target/lib/akka-osgi_2.11-sources.jar"/>
        <classpathentry kind="lib" path="target/lib/akka-stream_2.11.jar" sourcepath="target/lib/akka-stream_2.11-sources.jar"/>
        <classpathentry kind="lib" path="target/lib/client-all-2-11.jar" sourcepath="target/lib/client-all-2-11-sources.jar"/>
        <classpathentry kind="output" path="target/classes"/>
```

File: org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ScalaProject.scala (hunk 5 above)
Comment (sschaef) [inline org.scala-ide.sdt.core/src/org/scalaide/core/internal/project/ScalaProject.scala]: this comment is not yet resolved.

**Comment by sschaef** [comment]
The test runs fine on my system, but I got an exception on shutdown:
```
!ENTRY org.eclipse.core.resources 2 2 2016-08-08 20:19:54.690
!MESSAGE Save operation warnings.
!SUBENTRY 1 org.eclipse.core.resources 2 566 2016-08-08 20:19:54.690
!MESSAGE Problems occurred during save.
!STACK 0
java.lang.IllegalStateException: BundleContext is no longer valid
	at org.eclipse.osgi.internal.framework.BundleContextImpl.checkValid(BundleContextImpl.java:984)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getServiceReference(BundleContextImpl.java:559)
	at org.eclipse.pde.internal.core.PDECore.acquireService(PDECore.java:359)
	at org.eclipse.pde.internal.core.target.P2TargetUtils.cleanOrphanedTargetDefinitionProfiles(P2TargetUtils.java:173)
	at org.eclipse.pde.internal.core.PDECore$1.saving(PDECore.java:283)
	at org.eclipse.core.internal.resources.SaveManager.executeLifecycle(SaveManager.java:389)
	at org.eclipse.core.internal.resources.SaveManager$1.run(SaveManager.java:198)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.resources.SaveManager.broadcastLifecycle(SaveManager.java:201)
	at org.eclipse.core.internal.resources.SaveManager.save(SaveManager.java:1186)
	at org.eclipse.core.internal.resources.SaveManager.save(SaveManager.java:1165)
	at org.eclipse.core.internal.resources.DelayedSnapshotJob.run(DelayedSnapshotJob.java:52)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
```

Did you also see this one? I wonder why it complains about not being able to save.

**Comment by sschaef** [comment]
I did a few behavior tests in some example projects and so far it works really great. Good work Wieslaw! :+1:

Except of my minor remarks above, this LGTM.

**Pull request merged**